### PR TITLE
Unit tests for the Google.Cloud.Diagnostics.AspNetCore Error Reporting Exception Logger. 

### PR DESCRIPTION
**Pull request by iantalarico**

### CODE REVIEW

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/ErrorReporting/ErrorReportingExceptionLogger.cs [hunk 1]
```diff
@@ -145,8 +145,9 @@ private HttpRequestContext CreateHttpRequestContext(HttpContext context)
             return new HttpRequestContext()
             {
                 Method = request?.Method?.ToString() ?? "",
-                Url = UriHelper.GetDisplayUrl(request) ?? "",
-                UserAgent = request?.Headers["User-Agent"] ?? "",
+                Url = request?.GetDisplayUrl() ?? "",
+                UserAgent = string.IsNullOrEmpty(request?.Headers["User-Agent"]) ?
+                        "" : request.Headers["User-Agent"].ToString(),
```
Comment (evildour) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/ErrorReporting/ErrorReportingExceptionLogger.cs]: Why was this line changed? It seems like this is equivalent to the old logic.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/ErrorReporting/ErrorReportingExceptionLogger.cs [hunk 2]
```diff
@@ -163,8 +164,8 @@ private SourceLocation CreateSourceLocation(Exception exception)
             }
 
             StackTrace stackTrace = new StackTrace(exception, true);
-            StackFrame[] frames = stackTrace.GetFrames();
-            if (frames.Length == 0)
+            StackFrame[] frames = stackTrace?.GetFrames();
```
Comment (evildour) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/ErrorReporting/ErrorReportingExceptionLogger.cs]: This change seems unnecessary. `stackTrace` cannot be null since it's created on the previous line.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/ErrorReporting/ErrorReportingExceptionLoggerTest.cs [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Cloud.Diagnostics.AspNetCore.Tests.ErrorReporting
+{
+    public class ErrorReportingExceptionLoggerTest
+    {
+        private static readonly string _projectId = "pid";
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/ErrorReporting/ErrorReportingExceptionLoggerTest.cs]: Make the constants const? Otherwise, make the static field names `s_projectId` etc. (I don't much mind which.)

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/ErrorReporting/ErrorReportingExceptionLoggerTest.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            var context = new DefaultHttpContext();
+            context.Request.Host = new HostString(_googleHost);
+
+            mockClient.Setup(client => client.ReportErrorEvent(_projectName, IsSimpleContext(), null));
+            logger.Report(context, new Exception(_exceptionMessage));
+            mockClient.VerifyAll();
+        }
+
+        [Fact]
+        public void Report()
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.Tests/ErrorReporting/ErrorReportingExceptionLoggerTest.cs]: Report_Complex as a parallel to Report_Simple?

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/ErrorReporting/ErrorReportingExceptionLogger.cs (hunk 1 above)
Comment (iantalarico) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/ErrorReporting/ErrorReportingExceptionLogger.cs]: I updated this I didn't realize it was a StringValue, not a string.

File: apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/ErrorReporting/ErrorReportingExceptionLogger.cs [hunk 5]
```diff
@@ -145,8 +145,8 @@ private HttpRequestContext CreateHttpRequestContext(HttpContext context)
             return new HttpRequestContext()
             {
                 Method = request?.Method?.ToString() ?? "",
-                Url = UriHelper.GetDisplayUrl(request) ?? "",
-                UserAgent = request?.Headers["User-Agent"] ?? "",
+                Url = request?.GetDisplayUrl() ?? "",
+                UserAgent = request?.Headers["User-Agent"].ToString() ?? "",
```
Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/ErrorReporting/ErrorReportingExceptionLogger.cs]: So if these are StringValue values, do we need anything more than

```csharp
Url = request?.GetDisplayUrl(),
UserAgent = request?.Headers["User-Agent"]
```

Shouldn't we leave them assigning null if there's no value?

Comment (iantalarico) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/ErrorReporting/ErrorReportingExceptionLogger.cs]: The set for  UserAgent will throw if we try to assign it a null value so we either need to not set it or use the empty string.

Comment (jskeet) [inline apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/ErrorReporting/ErrorReportingExceptionLogger.cs]: Ah - they're not actually StringValue, they're string fields according to https://github.com/googleapis/googleapis/blob/master/google/devtools/clouderrorreporting/v1beta1/common.proto - that's why they'd throw. It's fine to set a StringValue field to null.

**Comment by iantalarico** [comment]
Oh I see I meant the Headers["User-Agent"] was a StringValue.  I think we were talking about two different things :)
